### PR TITLE
Add exec test

### PR DESCRIPTION
**Pull request by xlgao-zju**

### CODE REVIEW

File: pkg/validate/container.go [hunk 1]
```diff
@@ -64,137 +64,137 @@ var _ = framework.KubeDescribe("Container", func() {
...
-
-			By("check the log context")
-			expectedLogMessage := &logMessage{
-				log:    []byte(defaultLog + "\n"),
-				stream: stdoutType,
-			}
-			verifyLogContents(podConfig, logPath, expectedLogMessage)
-		})
-	})
+	// Context("runtime should support basic operations on container", func() {
```
Comment (feiskyer) [inline pkg/validate/container.go]: Hint: you don't need to remove those codes, using `ginkgo` flags instead, e.g. `critest --ginkgo-flags=-focus=Streaming`

Comment (xlgao-zju) [inline pkg/validate/container.go]: @feiskyer Thank you, it's very helpful.

Comment (xlgao-zju) [inline pkg/validate/container.go]: have got other tests back.

**Comment by feiskyer** [comment]
@xlgao-zju Check https://github.com/feiskyer/cri-tools/commit/803c07516e44daa5a3b735395be225d86e5a8a3d#diff-d1f71bbe5606eb567619242dad1d9797R142 for methods of visiting the streaming endpoints.

**Comment by xlgao-zju** [comment]
@feiskyer Updated.

File: pkg/validate/streaming.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+func checkExec(c internalapi.RuntimeService, req string) {
+	var streamOut, streamErr io.Writer
+	localOut := &bytes.Buffer{}
+	localErr := &bytes.Buffer{}
+
+	streamOut = localOut
+	streamErr = localErr
+
+	url, err := url.Parse(defaultStreamServer + req)
```
Comment (feiskyer) [inline pkg/validate/streaming.go]: Should check whether the streaming URL is full URL, it should be a full URL for external runtimes, but dockershim only returns a relative path.

File: pkg/validate/streaming.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	framework.ExpectNoError(err, "failed to parse url:  %q", defaultStreamServer+req)
+	conf := &rest.Config{
+		Host: defaultStreamServer,
+	}
+
+	e, err := remotecommand.NewExecutor(conf, "GET", url)
```
Comment (feiskyer) [inline pkg/validate/streaming.go]: should be "POST"

**Comment by xlgao-zju** [comment]
@feiskyer updated

File: pkg/validate/streaming.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validate
+
+import (
+	"bytes"
+	//"io"
```
Comment (feiskyer) [inline pkg/validate/streaming.go]: nit: remove unused imports

File: pkg/validate/streaming.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+	url := parseURL(c, req)
+	conf := &rest.Config{
+		Host: url.Host,
+	}
+	e, err := remotecommand.NewExecutor(conf, "POST", url)
+	framework.ExpectNoError(err, "failed to create executor for %q", req)
+
+	err = e.Stream(remotecommand.StreamOptions{
+		SupportedProtocols: []string{remote.StreamProtocolV1Name},
```
Comment (feiskyer) [inline pkg/validate/streaming.go]: should be `remotecommandconsts.SupportedStreamingProtocols`

File: pkg/validate/streaming.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	framework.Logf("Get exec url: " + resp.Url)
+	return resp.Url
+}
+
+func checkExec(c internalapi.RuntimeService, req string) {
+	localOut := &bytes.Buffer{}
+	localErr := &bytes.Buffer{}
+
+	url := parseURL(c, req)
+	conf := &rest.Config{
```
Comment (feiskyer) [inline pkg/validate/streaming.go]: Add a comment and todo for tls.

// Only http is supported now.
// TODO: support streaming APIs via tls.

File: pkg/validate/streaming.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+	Expect(localOut.String()).To(Equal("hello\n"), "The stdout of exec should be hello")
+	Expect(localErr.String()).To(BeEmpty(), "The stderr of exec should be empty")
+	framework.Logf("Check exec url %q succeed", req)
+}
+
+func parseURL(c internalapi.RuntimeService, req string) *url.URL {
+	version := getVersion(c)
+	if version.RuntimeName == "docker" {
+		req = defaultStreamServer + req
```
Comment (feiskyer) [inline pkg/validate/streaming.go]: It's better to make this robust by URL itself, e.g. only concat if scheme and host is not specified in the URL.

File: pkg/validate/streaming.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+		Tty:                false,
+	})
+	framework.ExpectNoError(err, "failed to open streamer for %q", req)
+
+	Expect(localOut.String()).To(Equal("hello\n"), "The stdout of exec should be hello")
+	Expect(localErr.String()).To(BeEmpty(), "The stderr of exec should be empty")
+	framework.Logf("Check exec url %q succeed", req)
+}
+
+func parseURL(c internalapi.RuntimeService, req string) *url.URL {
```
Comment (feiskyer) [inline pkg/validate/streaming.go]: rename `req` to `execServerURL` for clear

File: pkg/validate/streaming.go (hunk 7 above)
Comment (xlgao-zju) [inline pkg/validate/streaming.go]: But dockershim will only return url like `/cri/exec/token`, right? So if I did not miss anything , is it necessary to check  scheme and host?

Comment (xlgao-zju) [inline pkg/validate/streaming.go]: @feiskyer

Comment (feiskyer) [inline pkg/validate/streaming.go]: What if it changed later?

Comment (xlgao-zju) [inline pkg/validate/streaming.go]: ok, will do it for the future.

File: pkg/validate/streaming.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	framework.ExpectNoError(err, "failed to parse url:  %q", execServerURL)
+
+	version := getVersion(c)
+	if version.RuntimeName == "docker" {
+		if url.Host == "" {
+			url.Host = defaultStreamServerAddress
+		}
+		if url.Scheme == "" {
+			url.Scheme = defaultStreamServerScheme
+		}
```
Comment (xlgao-zju) [inline pkg/validate/streaming.go]: @feiskyer fixed.

**Comment by feiskyer** [comment]
LGTM

**Pull request merged**